Automattic/jetpack
# Limit amount of data send by full sync

## Proposed changes:

We are seeing many OOMs caused by **Full Sync** which tries to send as much data as possible, ignoring the `max_upload_size` restriction, due to the nature how Jetpack Sync queues items. 

This is an approach to send smaller chunks when in **Full Sync** if they are bigger than a specific threshold only for **Post objects** since they are the ones that feel like might be a problem based on size. 

Prior logic used to chunk Posts only based on defaults using `chunk_size` and `max_chunks`. This PR overrides the `get_next_chunk` method for the `Posts` Module and incorporates existing logic about expanding posts and adding metadata that was being done later down the road in the Full Sync Process. Also, metadata trimming has been added to the process since it did not exist for **Full Sync**.

The idea is to make sure we limit the max amount of posts and their metadata associated to a maximum size. (For the moment it is  `MAX_POST_META_LENGTH + MAX_POST_CONTENT_LENGTH`) All of it after trimming both posts and metadata.

In the edge case of the first post and its associated metadata being bigger than the max size allowed after trimming the content, we allow only that one post to be synced. Otherwise, we will never be able to sync that object. We might need to monitor that edge case to see if it still causes issues.



### Other information:


## Jetpack product discussion


pf5801-gq-p2


## Does this pull request change what data or activity we track or use?



## Testing instructions:






Let's start with a self-hosted site without this branch applied (be it your local or a JN)
* Create a bunch of posts (more than 100) with some big content. I suggest creating a very big one and using some plugin to easily duplicate
* Go to Jetpack Debugger and Schedule a Full Sync for posts only.
* Verify the content was synced by chunk sizes of 100 checking the corresponding sync-audit logs (link provided in the Jetpack Debugger but change the queue to `immediate-send`). 

Now move the site to this branch
* Go to Jetpack Debugger 
* Just in case, clean the Full Sync queue by scheduling a Full Sync for options, constants, functions and updates.
* Schedule a Full Sync for posts only.
* Verify the content was synced by smaller chunk sizes than 100 checking the corresponding sync-audit logs (link provided in the Jetpack Debugger but change the queue to `immediate-send`). The specific chunking size will depend on how big your posts are.
